### ISSUE TITLE
Feature request: Accept the *filter* on standard-input

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [stedolan/jq#1582]
In quite a lot of cases, the *filter* I want to write is a little longer, but I have no need of standard-input for the *input*, because it exists in a file. In such cases, I'd really like to use `stdin` for the jq-language code, instead of the input.

I propose the standard `-f -` syntax for this:

```sh
jq -f - package.json <<FILTER
   [.dependencies | to_entries | .[] | .key+"@"+.value]
      | join(", ")
FILTER
```

There are, I'm sure, concerns with how this would operate with the few I/O primitives that exist: perhaps disable them entirely when `-f` is using `stdin`? Or perhaps we could accept them on a different file-descriptor, `3` for instance? `:3`

**Comment by nicowilliams** [comment]
@ELLIOTTCABLE Yes, I agree.  You can actually get this on Unix like so right now:

```
$ jq -nf /dev/stdin
.
^D
null
$ 
```

But obviously it won't work on Windows.

This also works, on a tty only on Unix:

```
$ jq -f
.
^D1
^D
1
$ 
```

But yes, `jq -f -` should be equivalent to `jq -nf /dev/stdin` and I see no reason not to make it so.

**Comment by ELLIOTTCABLE** [comment]
`<3`!

**Comment by pkoppstein** [comment]
On Windows, CON: is the console, so one can use it in lieu of /dev/stdin to read from the console:

```
$ jq -f con: input.json
length
^Z
3
```
(Control-Z must be used for "end of input".)

Since CON: is not STDIN, we can have our cake and eat it:

```
$ jq -f con:
length
^Z
[0]
1
[0,1,2]
3
^Z
```

In a way, the same applies in other environments if bash-style process redirection works:

```
jq -f <(cat<<EOF
length
EOF
)
```

An example:
```
$ echo length | jq -R -f <(cat) <<< abc
3
```

**Comment by dkrieger** [comment]
I second the process redirection approach instead of adding this as a feature. If we're talking useful syntactic sugar that doesn't invert the basic interface, I'd love the ability to add multiple files using '-f' as well as combining a file with the final string argument. The library enhancement issue was closed without apparent resolution, and simply allowing multiple file inputs (combined with string filter for bonus points) would vastly improve the jq interface.

**Comment by pkoppstein** [comment]
@dkrieger - Maybe some of these possible enhancements each deserve separate "Enhancement Request" tickets.  

One of the needed enhancements, it seems to me, follows from the disappearance of support for the automatic inclusion of a file (~/.jq) when ~/.jq is a directory.  It would be an easy enhancement to have jq read in a file (say ~/.jq/.jq or ~/.jq/jqrc) if ~/.jq is a directory.  

This is relevant in the present context because the jq maintainers are extremely reluctant to add command-line options, as would I think be required to support the likes of `jq -f FILENAME -e STRING JSONFILE ...`

What is the "library enhancement issue" you referenced?

**Comment by dkrieger** [comment]
@pkoppstein I was referring to https://github.com/stedolan/jq/issues/472

I am not opposed to using a home-based library directory, but IMO given the way I use jq I think I'd care more about including functions from a pwd-relative directory. It seems insufficient to be limited to a globally namespaced library when one might need (not-so-)little helper functions for a given task. The way I use jq is typically when I'm rapidly developing a solution that isn't quite throwaway but isn't quite "production quality" (much like bash scripts), so the ability to define helper functions local to the problem domain (typically contained within a directory) is essential. One could always have a global collection of functions arbitrarily defined in the home directory and symlink them into the pwd-relative library directory.

**Comment by nicowilliams** [comment]
Let's not conflate things.  This is just about the main program.

**Comment by izeau** [comment]
Here is a variation on @pkoppstein’s method with no dangling parenthesis using a command group:

```sh
{ jq -f <(cat) input.json; } <<'FILTER'
length
FILTER
```
